pkg/kubeapiserver/admission/configuration/configuration_manager.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configuration
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	interval  = 1 * time.Second
+	threshold = 5
+)
+
+type getFunc func() (runtime.Object, error)
+
+// When running, configurationManger calls `get` every `interval`. If `get` is
+// successful, `Ready()` returns ready and `configuration()` returns the
+// `mergedConfiguration`; if `get` has failed more than `threshold` times,
+// `Ready()` returns not ready and `configuration()` returns nil configuration.
+type configurationManager struct {
+	// a function to consistently read the latest configuration
+	get getFunc
+	// consistent read interval
+	interval time.Duration
+	// if the number of consecutive read failure equals or exceeds the threshold, the
+	// configuration is regarded as not ready.
+	threshold uint
+	// if the configuration is regarded as ready.
+	ready     bool
+	readyLock sync.RWMutex
+
+	mergedConfiguration runtime.Object
+	configurationLock   sync.RWMutex
+}
+
+func newAdmissionControlConfigurationManager(get getFunc) *configurationManager {
+	return &configurationManager{
+		get:       get,
+		interval:  interval,
+		threshold: threshold,
+	}
+}
+
+func (a *configurationManager) Ready() bool {
+	a.readyLock.RLock()
+	defer a.readyLock.RUnlock()
+	return a.ready
+}
+
+func (a *configurationManager) setReady(value bool) {
+	a.readyLock.Lock()
+	defer a.readyLock.Unlock()
+	a.ready = value
+}
+
+func (a *configurationManager) configuration() (runtime.Object, error) {
+	if !a.Ready() {
+		return nil, fmt.Errorf("configuration is not ready")
+	}
+	a.configurationLock.RLock()
+	defer a.configurationLock.RUnlock()
+	return a.mergedConfiguration, nil
+}
+
+func (a *configurationManager) setConfiguration(value runtime.Object) {
+	a.configurationLock.Lock()
+	defer a.configurationLock.Unlock()
+	a.mergedConfiguration = value
+}
+
+func (a *configurationManager) Run(stopCh <-chan struct{}) {
+	var failure uint
+	go wait.Until(func() {
+		configuration, err := a.get()
+		if err != nil {
+			failure++
+			if failure >= a.threshold {
+				a.setReady(false)
+			}
+			return
+		}
+		failure = 0
+		a.setConfiguration(configuration)
```

Comment:
I think setting in this order is probably OK but if you lock around both of them then it is obvious that there is not a problem.